File: src/BeepBong.Web.Vue/ClientApp/models/pagination.ts
export class Pagination {
    constructor()
    {
        this.pageIndex = 1;
        this.totalItems = 0;
        this.totalPages = 1;
    }

    pageIndex: number;
    totalItems: number;
    totalPages: number;

    // get hasNextPage(): boolean {
    //     if (this.pageIndex < 0) return false;
    //     return this.pageIndex < this.totalPages;
    // }

    // get hasPreviousPage(): boolean {
    //     if (this.pageIndex < 0) return false;
    //     return this.pageIndex > 1; 
    // }

    // get onFirstPage(): boolean {
    //     if (this.pageIndex < 0) return true;
    //     return this.pageIndex == 1; 
    // }

    // get onLastPage(): boolean {
    //     if (this.pageIndex < 0) return true;
    //     return this.pageIndex == this.totalPages; 
    // }
}

File: src/BeepBong.DataAccess/Pagination.cs
using Microsoft.EntityFrameworkCore;
using System.Threading.Tasks;
using System.Linq;
using System.Collections.Generic;
using System;

namespace BeepBong.DataAccess
{
    public class Pagination<T>
    {
        public int PageIndex { get; }
        public int TotalPages { get; }
        public int TotalItems { get; }
        public List<T> Items { get; } = new List<T>();

        public Pagination(List<T> items, int count, int pageIndex, int pageSize)
        {
            PageIndex = pageIndex;
            TotalPages = (int)Math.Ceiling(count / (double)pageSize);
            TotalItems = count;

            this.Items.AddRange(items);
        }

        // public bool HasPreviousPage
        // {
        //     get => PageIndex > 1;
        // }

        // public bool HasNextPage
        // {
        //     get => PageIndex < TotalPages;
        // }

        public static async Task<Pagination<T>> CreateAsync(IQueryable<T> source, int pageIndex, int pageSize)
        {
            var count = await source.CountAsync();
            var items = await source.Skip((pageIndex - 1) * pageSize).Take(pageSize).ToListAsync();
            return new Pagination<T>(items, count, pageIndex, pageSize);
        }
    }
}